website/blog/2023-02-10.ferretdb-compare-arrays.md

```diff
@@ -0,0 +1,504 @@
+* `{ _id: 'array-empty', v: [] }` is selected as empty array is the smallest.
+* `{ _id: 'array-mixed', v: [ 0, 'a', 'z' ] }` is selected by comparing 
+  each element and first element `0` is less than `c` in comparison order.
+* `{ _id: 'array-numbers', v: [ 10, 5 ] }` is selected by comparing
+  each element and first element `0` is less than `c` in comparison order.
+* `{ _id: 'array-strings', v: [ 'd' ] }` is not selected because comparing 
+  the first element `d` is less not than `c`.
+* `{ _id: 'string', v: 'b' }` and `{ _id: 'null', v: null }` are not selected,
+  they are not arrays.
+
+```js
+db.mixedfilter.find({v:{$lt:['c']}})
+[
+  { _id: 'array-empty', v: [] },              // empty array is the smallest array
+  { _id: 'array-mixed', v: [ 0, 'a', 'z' ] }, // compares each element of [ 0, 'a', 'z' ]
+  { _id: 'array-numbers', v: [ 10, 5 ] }      // compares each element of [ 10, 5 ]
+]
+```
+
+Let's look at filtering greater than `['c']`.
+It selects array BSON type and compares the array against `['c']`.
+```js
+db.mixedfilter.find({v: {$gt: ['c']}})
+[ { _id: 'array-strings', v: [ 'd' ] } ] // compares each element of [ 'd' ]
+```
+#### Filtering nested array 
+Let's look at filtering nested array with an array.
+```js
+db.nestedfilter.insertMany([
+  { _id: 'nested-empty', v: [ [] ] },
+  { _id: 'nested-mixed', v: [ [ 'a', 1 ] ] },
+  { _id: 'nested-string', v: [ [ 'g', 'a' ] ] },
+  { _id: 'nested-number', v: [ [ 10, 1 ] ] },
+  { _id: 'array-bool', v: [ true ] },
+  { _id: 'array-null', v: [ null ] }
+])
+```
+Using non-array filter such as `'c'`, nothing was matched since `$gt` and `$lt`
+does not find string BSON type within the array, they are in the nested array.
+```js
+db.nestedfilter.find({v:{$lt: 'c'}})
+
+db.nestedfilter.find({v:{$gt: 'c'}})
+```
+
+##### $lt
+Using array filter `['c']` on `$lt` operator, it selected 
+empty nested array `{ _id: 'nested-empty', v: [ [] ] }`
+and `{ _id: 'nested-mixed', v: [ [ 'a', 1 ] ] }` which the nested array contains
+smaller element `a` than filter `c` on its first index.
+* selected `{ _id: 'nested-empty', v: [ [] ] }` because empty nested array is less.
+* selected `{ _id: 'nested-mixed', v: [ [ 'a', 1 ] ] }` because first index of 
+  nested array `a` is less than `c`.
+* selected ` { _id: 'nested-number', v: [ [ 10, 1 ] ] }` because first index of nested array
+ `10` is less than `c` in BSON comparison order.
+* selected `{ _id: 'array-null', v: [ null ] }` because `null` less than string `c`
+  in BSON comparison order.
+* not selecting `{ _id: 'nested-string', v: [ [ 'g', 'a' ] ] }` because `g` is not
+  less than `c`.
+* not selecting `{ _id: 'array-bool', v: [ true ] }` because boolean is not less than
+  string in BSON comparison order.
+```js
+db.nestedfilter.find({v:{$lt: ['c']}})
+[
+  { _id: 'nested-empty', v: [ [] ] },
+  { _id: 'nested-mixed', v: [ [ 'a', 1 ] ] },
+  { _id: 'nested-number', v: [ [ 10, 1 ] ] },
+  { _id: 'array-null', v: [ null ] }
+]
```

Comment:
@b1ron I felt I was pretty sure I understood this. But looking at nested array with $gt below, I'm not so sure that my interpretation is correct anymore. Could you have a look please :pray:

Reply:
I deleted my review comments to avoid mass confusion. I think it's due to your query shape. It should be `[ [ 'c' ] ]` to match nested elements with match expression. The other behavior seems to be undefined.

```
> db.nestedfilter.find({ v:{ $lt: [ [ 'c' ] ]}})
{ "_id" : "nested-empty", "v" : [ [ ] ] }
{ "_id" : "nested-mixed", "v" : [ [ "a", 1 ] ] }
{ "_id" : "nested-string", "v" : [ [ "g", "a" ] ] }
{ "_id" : "nested-number", "v" : [ [ 10, 1 ] ] }
{ "_id" : "array-null", "v" : [ null ] }
{ "_id" : 1, "v" : [ [ "a" ] ] }
{ "_id" : 2, "v" : [ [ "z" ] ] }
{ "_id" : 3, "v" : [ [ "0" ] ] }
> 
```

With `$gt` match.

```
> db.nestedfilter.find({ v:{ $gt: [ [ 'c' ] ]}})
{ "_id" : "nested-string", "v" : [ [ "g", "a" ] ] }
{ "_id" : "array-bool", "v" : [ true ] }
{ "_id" : 2, "v" : [ [ "z" ] ] }
```

Reply:
Actually `$lt` now seems to be return incorrect results, _id 2 is GT.

Reply:
Yes I tried, `[['c']]` as well. But I decided that this is not as interesting as `['c']`  two nested array _should_ just compare by iterating each index. So that _should_ be simple. But it's not since `$lt` behaves weird, which was surprising.

So query shape of `['c']` instead of `[['c']]` was on purpose. Because we want to understand using non-array filte, and array filter on nested arrays. :hugs:

Reply:
I can remove some examples, if they are undefined :hugs: